akka-remote/src/main/scala/akka/remote/artery/tcp/ArteryTcpTransport.scala

```diff
@@ -0,0 +1,405 @@
+/**
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.remote.artery
+package tcp
+
+import java.net.InetSocketAddress
+
+import scala.concurrent.Await
+import scala.concurrent.ExecutionContext
+import scala.concurrent.ExecutionContextExecutor
+import scala.concurrent.Future
+import scala.concurrent.Promise
+import scala.concurrent.duration._
+import scala.util.Failure
+import scala.util.Success
+import scala.util.Try
+import akka.ConfigurationException
+import akka.Done
+import akka.NotUsed
+import akka.actor.ActorSystem
+import akka.actor.ExtendedActorSystem
+import akka.dispatch.ExecutionContexts
+import akka.event.Logging
+import akka.remote.RemoteActorRefProvider
+import akka.remote.RemoteTransportException
+import akka.remote.artery.Decoder.InboundCompressionAccess
+import akka.remote.artery.compress._
+import akka.stream.Attributes
+import akka.stream.Attributes.LogLevels
+import akka.stream.FlowShape
+import akka.stream.Graph
+import akka.stream.KillSwitches
+import akka.stream.Materializer
+import akka.stream.SharedKillSwitch
+import akka.stream.SinkShape
+import akka.stream.scaladsl.Broadcast
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.GraphDSL
+import akka.stream.scaladsl.Keep
+import akka.stream.scaladsl.MergeHub
+import akka.stream.scaladsl.Partition
+import akka.stream.scaladsl.RestartFlow
+import akka.stream.scaladsl.Sink
+import akka.stream.scaladsl.Source
+import akka.stream.scaladsl.Tcp
+import akka.stream.scaladsl.Tcp.IncomingConnection
+import akka.stream.scaladsl.Tcp.ServerBinding
+import akka.util.ByteString
+import akka.util.OptionVal
+
+/**
+ * INTERNAL API
+ */
+private[remote] object ArteryTcpTransport {
+
+  private val successUnit = Success(())
+
+  def optionToTry(opt: Option[Throwable]): Try[Unit] = opt match {
+    case None    ⇒ successUnit
+    case Some(t) ⇒ Failure(t)
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] class ArteryTcpTransport(_system: ExtendedActorSystem, _provider: RemoteActorRefProvider)
+  extends ArteryTransport(_system, _provider) {
+  import ArteryTransport._
+  import ArteryTcpTransport._
+  import FlightRecorderEvents._
+
+  // may change when inbound streams are restarted
+  @volatile private var inboundKillSwitch: SharedKillSwitch = KillSwitches.shared("inboundKillSwitch")
```

Comment:
How are race conditions wrt to changing this field resolved? We don't control the time when new connections come in, so shouldn't it be possible that a new stream just comes in after `inboundKillSwitch.shutdown()` was called but before the new switch was set? Maybe that one can be avoided by first creating a new kill switch before shutting down the old one?

Reply:
If such connection is materialized after `inboundKillSwitch.shutdown()` it will fail immediately, which is fine. It's same result as if it happened a few millis earlier.

Reply:
Ah, I see.